### PR TITLE
HHH-15725 Criteria API Expression.as adds cast even when the cast type is equal to the expression type

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-15725 
This is an alternative solution to https://github.com/hibernate/hibernate-orm/pull/5634

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/AbstractCriteriaTest.java [hunk 1]
```diff
@@ -50,6 +50,8 @@
 		Thing.class,
 		ThingWithQuantity.class,
 		VersionedEntity.class
-})
+},
+		useCollectingStatementInspector = true
+)
```
Comment (mbladel) [inline hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/AbstractCriteriaTest.java:55]: Don't think you need this change anymore

File: hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AbstractSqmExpression.java [hunk 2]
```diff
@@ -94,6 +94,15 @@ public SqmExpression<String> asString() {
...
 	public <X> SqmExpression<X> as(Class<X> type) {
+		if ( getExpressible() != null ) {
+			final BasicType<X> basicTypeForJavaType = nodeBuilder().getTypeConfiguration()
+					.getBasicTypeForJavaType( type );
+			if ( basicTypeForJavaType != null
+					&& getExpressible().getRelationalJavaType().getJavaTypeClass()
+					== basicTypeForJavaType.getRelationalJavaType().getJavaTypeClass() ) {
+				return new AsWrapperSqmExpression<>( this, type );
+			}
+		}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AbstractSqmExpression.java:105]:
```suggestion
		final SqmExpressible<?> expressible = getExpressible();
		if ( expressible != null ) {
			if ( expressible.getExpressibleJavaType().getJavaTypeClass() == type ) {
				return this;
			}
			final BasicType<X> basicTypeForJavaType = nodeBuilder().getTypeConfiguration()
					.getBasicTypeForJavaType( type );
			if ( basicTypeForJavaType == null ) {
				throw new IllegalArgumentException( "Can't cast expression to unknown type: " + type.getCanonicalName() );
			}
			if ( expressible.getRelationalJavaType().getJavaTypeClass()
					== basicTypeForJavaType.getRelationalJavaType().getJavaTypeClass() ) {
				return new AsWrapperSqmExpression<>( basicTypeForJavaType, this );
			}
		}
```

File: hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AsWrapperSqmExpression.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+	public AsWrapperSqmExpression(SqmExpression<?> expression, Class<T> type) {
+		super(
+				expression.nodeBuilder().getTypeConfiguration().getBasicTypeForJavaType( type ),
+				expression.nodeBuilder()
+		);
+		this.expression = expression;
+	}
+
+	AsWrapperSqmExpression(SqmExpressible<T> type, SqmExpression<?> expression) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AsWrapperSqmExpression.java:25]:
```suggestion
	public AsWrapperSqmExpression(SqmExpressible<T> type, SqmExpression<?> expression) {
```

File: hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AsWrapperSqmExpression.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+	@Override
+	public <X> X accept(SemanticQueryWalker<X> walker) {
+		return walker.visitAsWrapperExpression( this );
+	}
+
+	@Override
+	public void appendHqlString(StringBuilder sb) {
+		expression.appendHqlString( sb );
+	}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AsWrapperSqmExpression.java:38]: We have to decide on a HQL representation for this expression. I propose `wrap(.. as ..)`

```suggestion
		sb.append( "wrap(" );
		expression.appendHqlString( sb );
		sb.append( " as " );
		sb.append( getNodeType().getReturnedClassName() );
		sb.append( ")" );
	}
	
	@Override
	public <X> SqmExpression<X> as(Class<X> type) {
		return expression.as( type );
	}
```

**Comment by gavinking** [comment]
The correct implementation is probably something like:

```java
	@Override @SuppressWarnings("unchecked")
	public <X> SqmExpression<X> as(Class<X> type) {
		return (SqmExpression<X>) this;
	}

```

**Comment by beikov** [comment]
> returning a **new** expression object

Thanks for pointing that out. I guess we should always return a new `AsWrapperSqmExpression` then to comply with the spec? Though that brings up the question of backwards compatibility... Casting was done ever since ORM 3.6. Also see https://github.com/hibernate/hibernate-orm/commit/eb0396b1c69e1cf069093df7774e60c0b609beb6#diff-44712025cd21f3fd778471c03a92e74170e252025ef123f56fc00b7d0c8d362bR45

**Comment by beikov** [comment]
Side note, [EclipseLink](https://github.com/eclipse-ee4j/eclipselink/blob/5b399938cf4bb4ec0856b2f658c3eff97e6f12fd/jpa/org.eclipse.persistence.jpa/src/main/java/org/eclipse/persistence/internal/jpa/querydef/ExpressionImpl.java#L74) always returns the same instance again and just does an unsafe type cast, which kind of also goes against the **new** part of that specification. I guess the spec should be clarified. Also, the TCK does not have a test for the `as()` method, so it's hard to say for sure what the intention was.

**Comment by gavinking** [comment]
@beikov Yeah, I agree, that's pretty weird, especially since IIRC Mike wrote that Javadoc.

OTOH, "returning a new expression object" is pretty explicit to me&mdash;I would say no clarification is needed.

**Comment by dreab8** [comment]
superseded by https://github.com/hibernate/hibernate-orm/pull/8681

**Pull request closed without merging**